### ISSUE TITLE
BC break: 2.7.2 => v2.8.0: Proxy generation: embedded documents can no longer be marked final

### ISSUE THREAD

**Issue by Bilge** [doctrine/common#813]
:warning: Full disclosure: I use `doctrine/mongodb`, not the One True Reference Implementation, ORM; in case that makes a difference.

Although it has never been possible to mark "root" documents `final`, embedded documents may be marked `final` since they are embedded inside other classes when generating proxies. Since 2.8.0, however, the following error is thrown where `Foo` is an `@ODM\EmbeddedDocument` class.

>[Doctrine\Common\Proxy\Exception\InvalidArgumentException]                   
  Unable to create a proxy for a final class "Foo".

**Comment by alcaeus** [comment]
Are you sure this worked for doctrine/common 2.7? ODM has always generated proxies for embedded documents (which I'm not sure makes sense), so I'd expect this would fail for 2.7 as well?

**Comment by Bilge** [comment]
@alcaeus The only thing I'm sure of is that it worked before this `composer install` (non-Doctrine projects omitted for brevity):

```
  - Updating doctrine/cache (v1.6.1 => v1.7.0): Downloading (100%)
  - Updating doctrine/collections (v1.4.0 => v1.5.0): Downloading (100%)
  - Updating doctrine/inflector (v1.1.0 => v1.2.0): Downloading (100%)
  - Updating alcaeus/mongo-php-adapter (1.1.1 => 1.1.2): Downloading (100%)
  - Updating doctrine/annotations (v1.4.0 => v1.5.0): Downloading (100%)
  - Updating doctrine/common (v2.7.2 => v2.8.0): Downloading (100%)
  - Updating doctrine/mongodb (1.5.0 => 1.6.0): Downloading (100%)
```

**Comment by alcaeus** [comment]
That's very strange. Could you provide a trace to the error above? I'd like to see what's responsible for trying to create a proxy for an embedded document. Couldn't find anything at first try.

**Comment by Bilge** [comment]
Sorry but an image is the best I can do...

![proxy generation stack trace](https://user-images.githubusercontent.com/470626/29208507-49a95ad2-7e82-11e7-9dc3-9a7f97282733.png)

**Comment by alcaeus** [comment]
Thanks for the trace I now see what's going on.

https://github.com/doctrine/common/pull/771 introduced a sanity check to ensure no invalid classes will be generated. I wouldn't call that a BC break, since instantiating such a proxy object would always cause a fatal error.

The fix should be ODM not generating proxy objects for embedded documents, since they won't be needed, ever. I wouldn't revert the change here, as the issue is purely within ODM. I've created https://github.com/doctrine/mongodb-odm/issues/1627 to track this, which will be fixed in 1.1.6. Until then I'd have to ask you to make your embedded documents non-final until we properly fix the issue.

**Comment by Bilge** [comment]
Thanks @alcaeus! We've already made those documents non-final, since we've learned it's usually impractical to wait for the next Doctrine release ;)
